server/src/main/java/org/elasticsearch/index/engine/TranslogOperationAsserter.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.elasticsearch.index.mapper.DocumentParser;
+import org.elasticsearch.index.mapper.MappingLookup;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ *
+ * A utility class to assert that translog operations with the same sequence number
+ * in the same generation are either identical or equivalent when synthetic sources are used.
+ */
+public abstract class TranslogOperationAsserter {
+    public static final TranslogOperationAsserter DEFAULT = new TranslogOperationAsserter() {
+    };
+
+    private TranslogOperationAsserter() {
+
+    }
+
+    public static TranslogOperationAsserter withEngineConfig(EngineConfig engineConfig) {
+        return new TranslogOperationAsserter() {
+            @Override
+            boolean sameIndexOperation(Translog.Index o1, Translog.Index o2) throws IOException {
+                if (super.sameIndexOperation(o1, o2)) {
+                    return true;
+                }
+                if (engineConfig.getIndexSettings().isRecoverySourceSyntheticEnabled()) {
+                    return super.sameIndexOperation(synthesizeSource(engineConfig, o1), o2)
+                        || super.sameIndexOperation(o1, synthesizeSource(engineConfig, o2));
+                }
+                return false;
+            }
+        };
+    }
+
+    static Translog.Index synthesizeSource(EngineConfig engineConfig, Translog.Index op) throws IOException {
+        final ShardId shardId = engineConfig.getShardId();
+        MappingLookup mappingLookup = engineConfig.getMapperService().mappingLookup();
+        DocumentParser documentParser = engineConfig.getMapperService().documentParser();
+        try (var reader = new TranslogDirectoryReader(shardId, op, mappingLookup, documentParser, engineConfig, () -> {})) {
```

Comment:
It is ok, but as an optional nit, I found `engineConfig` quite a big pass of information to the reader - since it only needs the analyzer and code IIUC. It confused me for a short while, thinking it needed more access. I'd prefer to pass the two specific params, but consider this optional given that it piggy-backs on existing work and the extension here is for assertions only.

Reply:
I've kept the EngineConfig parameter unchanged because we also need MapperService and Codec.